File: backend/app/api/contacts.py
"""Contacts API.

Org-scoped CRUD over the people an org reaches out to. A contact is a ``name``,
an ``email``, an optional linked ``account_id``, and a free-text ``role``. These
back the email recipient picker (Execute panel and chat) so outreach can target
a saved person by name instead of retyping an address.

Every endpoint resolves the caller's org via ``current_org`` and delegates to the
contacts repository, which is database-backed when a pool is configured and
process-local in-memory otherwise (so the offline demo and tests round-trip).
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, EmailStr, Field

from app.api._org import current_org
from app.repositories import contacts as contacts_repo

router = APIRouter(tags=["contacts"])


class ContactIn(BaseModel):
    """Payload to create a contact in the caller's org."""

    name: str = Field(description="Person's display name.")
    email: EmailStr = Field(description="Email address to reach the contact.")
    account_id: Optional[str] = Field(
        default=None, description="Optional account this contact belongs to."
    )
    role: Optional[str] = Field(
        default=None, description="Free-text role, e.g. 'Economic buyer'."
    )


class ContactUpdate(BaseModel):
    """Partial update for an existing contact (only set fields are applied)."""

    name: Optional[str] = None
    email: Optional[EmailStr] = None
    account_id: Optional[str] = None
    role: Optional[str] = None


@router.get("/contacts")
async def list_contacts(org_id: str = Depends(current_org)) -> List[Dict[str, Any]]:
    """Return every contact for the caller's org, most recent first."""

    return await contacts_repo.list_for_org(org_id)


@router.post("/contacts", status_code=201)
async def create_contact(
    body: ContactIn, org_id: str = Depends(current_org)
) -> Dict[str, Any]:
    """Create a contact in the caller's org."""

    if not (body.name or "").strip():
        raise HTTPException(status_code=400, detail="name must not be empty")
    return await contacts_repo.upsert(
        org_id,
        name=body.name.strip(),
        email=str(body.email),
        account_id=body.account_id,
        role=body.role,
    )


@router.put("/contacts/{contact_id}")
async def update_contact(
    contact_id: str,
    body: ContactUpdate,
    org_id: str = Depends(current_org),
) -> Dict[str, Any]:
    """Update fields on an existing contact in the caller's org."""

    existing = await contacts_repo.get(org_id, contact_id)
    if existing is None:
        raise HTTPException(status_code=404, detail="contact not found")

    fields = body.model_dump(exclude_unset=True)
    merged = dict(existing)
    for key, value in fields.items():
        merged[key] = str(value) if key == "email" and value is not None else value

    return await contacts_repo.upsert(
        org_id,
        name=merged.get("name") or existing["name"],
        email=merged.get("email") or existing["email"],
        account_id=merged.get("account_id"),
        role=merged.get("role"),
        contact_id=contact_id,
    )


@router.delete("/contacts/{contact_id}")
async def delete_contact(
    contact_id: str, org_id: str = Depends(current_org)
) -> Dict[str, Any]:
    """Remove a contact from the caller's org."""

    removed = await contacts_repo.delete(org_id, contact_id)
    if not removed:
        raise HTTPException(status_code=404, detail="contact not found")
    return {"deleted": contact_id}


@router.get("/accounts/{account_id}/contacts")
async def list_account_contacts(
    account_id: str, org_id: str = Depends(current_org)
) -> List[Dict[str, Any]]:
    """Return the org's contacts linked to a single account."""

    return await contacts_repo.list_for_account(org_id, account_id)

File: frontend/lib/api/contacts.ts
// Typed client for org-scoped contacts: the people an org reaches out to.
//
// Backs the Contacts page and the email recipient picker in the Execute panel.
// Every call sends the nba_session cookie (credentials: "include") so the
// backend resolves the caller's org, plus the optional bearer token when
// configured. Reads degrade to an empty list when the backend is unreachable so
// the picker never dead-ends; mutations surface real errors so forms can react.

import { API_BASE, ApiError } from "@/lib/api";
import { authHeaders } from "@/lib/auth";

export interface Contact {
  id: string;
  org_id?: string;
  name: string;
  email: string;
  account_id?: string | null;
  role?: string | null;
  created_at?: string;
}

// Form payload mirroring the backend ContactIn / ContactUpdate models.
export interface ContactInput {
  name: string;
  email: string;
  account_id?: string | null;
  role?: string | null;
}

export type ContactUpdateInput = Partial<ContactInput>;

async function send<T>(
  path: string,
  method: "POST" | "PUT" | "DELETE",
  body?: unknown,
): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    method,
    headers: authHeaders(
      body === undefined
        ? { Accept: "application/json" }
        : { "Content-Type": "application/json", Accept: "application/json" },
    ),
    credentials: "include",
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  if (!res.ok) {
    throw new ApiError(`${method} ${path} failed (${res.status})`, res.status);
  }
  return (await res.json()) as T;
}

/** List every contact for the caller's org. Empty list when offline. */
export async function listContacts(signal?: AbortSignal): Promise<Contact[]> {
  try {
    const res = await fetch(`${API_BASE}/contacts`, {
      headers: authHeaders({ Accept: "application/json" }),
      credentials: "include",
      cache: "no-store",
      signal,
    });
    if (!res.ok) {
      throw new ApiError(`GET /contacts failed (${res.status})`, res.status);
    }
    return (await res.json()) as Contact[];
  } catch {
    return [];
  }
}

/** List the org's contacts linked to one account. Empty list when offline. */
export async function listAccountContacts(
  accountId: string,
  signal?: AbortSignal,
): Promise<Contact[]> {
  try {
    const res = await fetch(
      `${API_BASE}/accounts/${encodeURIComponent(accountId)}/contacts`,
      {
        headers: authHeaders({ Accept: "application/json" }),
        credentials: "include",
        cache: "no-store",
        signal,
      },
    );
    if (!res.ok) {
      throw new Error(`GET account contacts failed (${res.status})`);
    }
    return (await res.json()) as Contact[];
  } catch {
    return [];
  }
}

/** Create a contact in the caller's org. */
export function createContact(payload: ContactInput): Promise<Contact> {
  return send<Contact>("/contacts", "POST", payload);
}

/** Apply a partial update to an existing contact. */
export function updateContact(
  id: string,
  payload: ContactUpdateInput,
): Promise<Contact> {
  return send<Contact>(`/contacts/${encodeURIComponent(id)}`, "PUT", payload);
}

/** Remove a contact from the caller's org. */
export function deleteContact(id: string): Promise<{ deleted: string }> {
  return send<{ deleted: string }>(
    `/contacts/${encodeURIComponent(id)}`,
    "DELETE",
  );
}
